src/string.cr

```diff
@@ -971,7 +971,156 @@ class String
     end
   end
 
-  # Returns a new string where each character yielded to the given block
+  # Returns a new string where the first character is yielded to the given
+  # block and replaced by its return value.
+  #
+  # ```
+  # "hello".sub {|x| (x.ord + 1).chr } #=> "iello"
+  # "hello".sub { "hi" } #=> "hiello"
+  # ```
+  def sub(&block : Char -> _)
+    return self if empty?
```

Comment:
the behaviour also seems not to be fully consistent through all variants of the method, e.g. ` def sub(hash : Hash(Char, _))` doesn't seem to return self when empty.